gocd/admin/cleanup.py

```diff
@@ -0,0 +1,213 @@
+        if agent["agent_config_state"] != "Disabled":
+            x = botmaster_patch(url, '{"agent_config_state": "Disabled"}', version=7)
+            if x.status_code != 200:
+                print("Can't disable agent", url, x, x.content)
+                continue
+        if not only_disable:
+            botmaster_delete(url, version=7)
+
+
+def cleanup_cache():
+    for file in glob.glob("/srv/go-repository-cache/repo-*solv*"):
+        os.unlink(file)
+    for suffix in ["openSUSE:Maintenance:*", "*:Staging:adi:*", "repo-*"]:
+        for dir in glob.glob(f"/srv/go-repository-cache/{suffix}"):
+            shutil.rmtree(dir)
+    for root, dir, files in os.walk("/srv/go-repository-cache/"):
+        if root.endswith("/.cache"):
+            for file in files:
+                os.unlink(os.path.join(root, file))
+
+
+def remove_old_runs(cur, pipeline):
+    cur.execute(
+        """SELECT distinct p.id,p.label from pipelines p join stages s on s.pipelineid = p.id
+                  and s.createdtime < current_date - interval '5' day and p.name=%s""",
+        (pipeline,),
+    )
+    ids = []
+    for row in cur.fetchall():
+        id, label = row
+        print("Remove", pipeline, label)
+        path = f"/var/lib/go-server/artifacts/pipelines/{pipeline}/{label}"
+        try:
+            shutil.rmtree(path)
+        except FileNotFoundError:
+            pass
+        ids.append(id)
+    cur.execute(
+        "delete from BUILDSTATETRANSITIONS where STAGEID in (select id from stages where pipelineid = ANY(%s))",
+        (ids,),
+    )
+    cur.execute(
+        "delete from PIPELINEMATERIALREVISIONS where pipelineid = ANY(%s)", (ids,)
+    )
+    cur.execute(
+        "delete from builds where stageid in (select id from stages where pipelineid = ANY(%s))",
+        (ids,),
+    )
+    cur.execute("delete from stages where pipelineid = ANY(%s)", (ids,))
+    cur.execute("delete from pipelines where id = ANY(%s)", (ids,))
+    cur.execute("update pipelines set NATURALORDER=0 where name = %s", (pipeline,))
+
+
+def cleanup_old_pipelines():
+    # very safe password - only reachable from localhost
+    conn = psycopg2.connect(
+        "dbname=gocd user=gocd_database_user password=gocd_database_password host=localhost"
+    )
+    with conn.cursor() as cur:
+        cur.execute("SELECT distinct name from pipelines")
+        pipelines = sorted([row[0] for row in cur.fetchall()])
+        for pipeline in pipelines:
+            remove_old_runs(cur, pipeline)
+            conn.commit()
+        cur.execute(
+            """delete from modifiedfiles where modificationid in
+               (select id from modifications where modifiedtime < current_date - interval '10' day)"""
+        )
+        # cur.execute("delete from modifications where modifiedtime < current_date - interval '10' day")
+    conn.close()
+
+
+def main():
+    # pull new image - if registry is down, better stop here
+    subprocess.run(
```

Comment:
Can you tag the old image?

Reply:
I can't, but I put the script in git so you can :)

Reply:
I can't test it, what about this:

```
   subprocess.run(
        [
            "docker",
            "tag",
            "registry.opensuse.org/opensuse/tools/images/containers_tumbleweed/gocd-agent-release-tools:latest",
            "registry.opensuse.org/opensuse/tools/images/containers_tumbleweed/gocd-agent-release-tools:previous",
        ],
        check=True,
    )

   subprocess.run(
        [
            "docker",
            "pull",
            "registry.opensuse.org/opensuse/tools/images/containers_tumbleweed/gocd-agent-release-tools:latest",
        ],
        check=True,
    )
```

Reply:
Tell me you don't love variables without telling me you don't love variables :)

Reply:
I added a variable to please all variable aficionados